servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/TransportObserverInitializer.java

```diff
@@ -16,46 +16,93 @@
 import io.netty.channel.ChannelPromise;
+import io.netty.handler.ssl.SslHandshakeCompletionEvent;
 
+import javax.net.ssl.SSLSession;
+
+import static io.servicetalk.transport.netty.internal.NettyPipelineSslUtils.extractSslSession;
+import static io.servicetalk.transport.netty.internal.TransportObserverUtils.assignConnectionError;
 import static io.servicetalk.transport.netty.internal.TransportObserverUtils.assignConnectionObserver;
+import static io.servicetalk.transport.netty.internal.TransportObserverUtils.securityHandshakeObserver;
 import static java.util.Objects.requireNonNull;
 
 /**
  * A {@link ChannelInitializer} that registers a {@link ConnectionObserver} for all channels.
  */
 public final class TransportObserverInitializer implements ChannelInitializer {
 
+    /**
+     * Tells which side is using secure connection.
+     */
+    public enum SecureSide {
+        CLIENT,
+        SERVER,
+        NONE
+    }
+
     private final TransportObserver transportObserver;
+    private final SecureSide secureSide;
 
     /**
      * Creates a new instance.
      *
      * @param transportObserver {@link TransportObserver} to initialize for the channel
+     * @param secureSide tells which side is using secure connection
      */
-    public TransportObserverInitializer(final TransportObserver transportObserver) {
+    public TransportObserverInitializer(final TransportObserver transportObserver, final SecureSide secureSide) {
         this.transportObserver = requireNonNull(transportObserver);
+        this.secureSide = secureSide;
     }
 
     @Override
     public void init(final Channel channel) {
         final ConnectionObserver observer = requireNonNull(transportObserver.onNewConnection());
         assignConnectionObserver(channel, observer);
-        channel.pipeline().addLast(new TransportObserverChannelHandler(observer));
+        final ChannelPipeline pipeline = channel.pipeline();
+        pipeline.addLast(new TransportObserverHandler(observer, secureSide));
     }
 
-    private static final class TransportObserverChannelHandler extends ChannelDuplexHandler {
+    private static final class TransportObserverHandler extends ChannelDuplexHandler {
         private final ConnectionObserver observer;
+        private final SecureSide secure;
+        private boolean handshakeStarted;
 
-        TransportObserverChannelHandler(final ConnectionObserver observer) {
+        TransportObserverHandler(final ConnectionObserver observer, final SecureSide secure) {
             this.observer = observer;
+            this.secure = secure;
+        }
+
+        @Override
+        public void handlerAdded(final ChannelHandlerContext ctx) {
+            if (secure != SecureSide.NONE && ctx.channel().isActive()) {
+                reportSecurityHandshakeStarting(ctx.channel());
+            }
+        }
+
+        @Override
+        public void channelActive(final ChannelHandlerContext ctx) {
+            if (secure == SecureSide.CLIENT) {
```

Comment:
Why is the condition in `handlerAdded` and `channelActive` different?
If this is valid, then add a comment to explain the same.

Reply:
That was my overthinking :) Channel is already active when server adds a handler. Therefore, there is no need to expect `channelActive` on the server-side. But we can make conditions the same with no side-effects. Thanks!